upstairs/src/upstairs.rs

```diff
@@ -0,0 +1,3370 @@
+                    up_jobs,
+                    ds_jobs,
+                    repair_done,
+                    repair_needed,
+                    extents_repaired: extents_repaired.to_vec(),
+                    extents_confirmed: extents_confirmed.to_vec(),
+                    extent_limit: extent_limit.to_vec(),
+                    live_repair_completed: live_repair_completed.to_vec(),
+                    live_repair_aborted: live_repair_aborted.to_vec(),
+                });
+                if r.is_err() {
+                    warn!(self.log, "control message reply failed");
+                }
+            }
+            ControlRequest::DownstairsWorkQueue(tx) => {
+                let out = self.downstairs.get_work_summary();
+                let r = tx.send(out);
+                if r.is_err() {
+                    warn!(self.log, "control message reply failed");
+                }
+            }
+        }
+    }
+
+    /// Checks if a repair is possible. If so, checks if any Downstairs is in
+    /// the [DsState::LiveRepairReady] state, indicating it needs to be
+    /// repaired. If a Downstairs needs to be repaired, try to start repairing
+    /// it. When starting the repair fails, this function will schedule a task
+    /// to retry the repair by setting [Self::repair_check_interval].
+    ///
+    /// If this Upstairs is [UpstairsConfig::read_only], this function will move
+    /// any Downstairs from [DsState::LiveRepairReady] back to [DsState::Active]
+    /// without actually performing any repair.
+    pub(crate) async fn on_repair_check(&mut self) -> RepairCheck {
+        info!(self.log, "Checking if live repair is needed");
+        if !matches!(self.state, UpstairsState::Active) {
+            info!(self.log, "inactive, no live repair needed");
+            self.repair_check_interval = None;
+            return RepairCheck::InvalidState;
+        }
+
+        if self.cfg.read_only {
+            info!(self.log, "read-only, no live repair needed");
+            // Repair can't happen on a read-only downstairs, so short circuit
+            // here. There's no state drift to repair anyway, this read-only
+            // Upstairs wouldn't have caused any modifications.
+            for c in self.downstairs.clients.iter_mut() {
+                c.skip_live_repair(&self.state);
+            }
+            self.repair_check_interval = None;
+            return RepairCheck::NoRepairNeeded;
+        }
+
+        // Verify that all downstairs and the upstairs are in the proper state
+        // before we begin a live repair.
+        let repair = self
+            .downstairs
+            .clients
+            .iter()
+            .filter(|c| c.state() == DsState::LiveRepair)
+            .count();
+
+        let repair_ready = self
+            .downstairs
+            .clients
+            .iter()
+            .filter(|c| c.state() == DsState::LiveRepairReady)
+            .count();
+
+        if repair > 0 {
```

Comment:
Got it, I've done that refactoring in 526993af034c7e2e1d32dd281eb72934e632e8cb

(that commit also required moving `create_test_upstairs` and `start_up_and_repair` into `upstairs.rs`, since they now touch `Upstairs` internals)